moshi/src/main/java/com/squareup/moshi/JsonWriter.kt

```diff
@@ -0,0 +1,548 @@
+
+  /**
+   * Sets whether object members are serialized when their value is null. This has no impact on
+   * array elements. The default is false.
+   *
+   * Returns true if object members are serialized when their value is null. This has no impact on
+   * array elements. The default is false.
+   */
+  @get:CheckReturnValue
+  public var serializeNulls: Boolean = false
+
+  @JvmField
+  public var promoteValueToName: Boolean = false
+
+  /**
+   * Controls the deepest stack size that has begin/end pairs flattened:
+   *
+   *  - If -1, no begin/end pairs are being suppressed.
+   *  - If positive, this is the deepest stack size whose begin/end pairs are eligible to be flattened.
+   *  - If negative, it is the bitwise inverse (~) of the deepest stack size whose begin/end pairs have been flattened.
+   *
+   *
+   * We differentiate between what layer would be flattened (positive) from what layer is being
+   * flattened (negative) so that we don't double-flatten.
+   *
+   *
+   * To accommodate nested flattening we require callers to track the previous state when they
+   * provide a new state. The previous state is returned from [beginFlatten] and restored
+   * with [endFlatten].
+   */
+  @JvmField
+  public var flattenStackSize: Int = -1
+
+  private var tags: MutableMap<Class<*>, Any>? = null
+
+  /** Returns the scope on the top of the stack.  */
+  public fun peekScope(): Int {
+    check(stackSize != 0) { "JsonWriter is closed." }
+    return scopes[stackSize - 1]
+  }
+
+  /** Before pushing a value on the stack this confirms that the stack has capacity.  */
+  public fun checkStack(): Boolean {
+    if (stackSize != scopes.size) return false
+    if (stackSize == 256) {
+      throw JsonDataException("Nesting too deep at $path: circular reference?")
+    }
+    scopes = scopes.copyOf(scopes.size * 2)
+    pathNames = pathNames.copyOf(pathNames.size * 2)
+    pathIndices = pathIndices.copyOf(pathIndices.size * 2)
+    if (this is JsonValueWriter) {
+      stack = stack.copyOf(stack.size * 2)
+    }
+    return true
+  }
+
+  public fun pushScope(newTop: Int) {
+    scopes[stackSize++] = newTop
+  }
+
+  /** Replace the value on the top of the stack with the given value.  */
+  public fun replaceTop(topOfStack: Int) {
+    scopes[stackSize - 1] = topOfStack
+  }
+
+  /**
+   * Sets the indentation string to be repeated for each level of indentation in the encoded
+   * document. If `indent.isEmpty()` the encoded document will be compact. Otherwise the
+   * encoded document will be more human-readable.
+   *
+   * @param indent a string containing only whitespace.
+   */
+  public open fun setIndent(indent: String) {
+    this.indent = indent.ifEmpty { null }
+  }
+
+  /**
+   * Returns a string containing only whitespace, used for each level of indentation. If empty, the
+   * encoded document will be compact.
+   */
+  @CheckReturnValue
+  public fun getIndent(): String {
+    return if (indent != null) indent!! else ""
```

Comment:
Please make this idiomatic (Elvis operator, `.orEmpty()`, etc)

Reply:
21c22db06e2430333a733a5e275cdf1eef9606d6